### ISSUE TITLE
Pipelining commands to the redis server: A ManyToMany use case with massive insertion of data

### ISSUE THREAD

**Issue by healiseu** [josiahcarlson/rom#130]
A fundamental part of my TRIADB project is a ManyToMany relationship between two nodes. The structure resembles that of a property graph, i.e. you define bidirectional edges that connect a tail node with an edge node and an edge node with the head node. You have outgoing edges, incoming edges, etc but you do not embed data values from records on the nodes, you link references of them as nodes (HyperAtoms) to other nodes (HyperBonds) that represent records (see this [hypergraph][3] for example). 

I managed to build this hypergraph of hyperedges (Entities) and hypernodes (Attributes) with ROM using two OneToMany relationships that are referencing an intermediary model (Hyperlink). Your example of building a many-to-many relationship with ROM helped me a lot. That works and I tested this on small sets but it takes ages to do bulk insertion of data. I have run an experiment to get an idea of what exactly I am talking about.

``` Python
class Airport(Model):
    pid     = Integer(unique=True, index=True)
    ident   = Text(unique=True, index=True, keygen=IDENTITY, prefix=True)
    type    = Text(index=True, keygen=IDENTITY, prefix=True)
    name    = Text(index=True, keygen=IDENTITY_CI, prefix=True)
    lat     = Float(index=True)
    long    = Float(index=True)
    elev    = Integer(index=True)
    cont    = Text(index=True, keygen=IDENTITY, prefix=True)
    cntr    = Text(index=True, keygen=IDENTITY, prefix=True)
    region  = Text(index=True, keygen=IDENTITY, prefix=True)
    municip = Text(index=True, keygen=IDENTITY_CI, prefix=True)
    serv    = Text(index=True, keygen=IDENTITY, prefix=True)
    gps     = Text(index=True, keygen=IDENTITY_CI, prefix=True)
    iata    = Text(index=True, keygen=IDENTITY, prefix=True)
    local   = Text(index=True, keygen=IDENTITY_CI, prefix=True)
    home    = Text(index=True, keygen=IDENTITY)
    wikip   = Text(index=True, keygen=IDENTITY)
    kwords  = Text(index=True, keygen=IDENTITY)
```
So first I ran a benchmark test on how long is going to take ROM to load on Redis a 1000rows x 18columns aiports dataset from a CSV file. Most of columns are fully filled
``` Python
"""
id                   1000
ident                1000
type                 1000
name                 1000
latitude_deg         1000
longitude_deg        1000
elevation_ft          989
continent               1
iso_country          1000
iso_region           1000
municipality         1000
scheduled_service    1000
gps_code             1000
iata_code               0
local_code           1000
home_link               0
wikipedia_link         16
keywords                7
"""
# ---------------------------------------- ROM ---------------------------------------------------------
# Time    : 2sec
# Stages                                  : CLEAN --> +loaded data
# info memory   - user_memory_dataset     : 59KB  -->     2.4MB (1MB .rdb)
# info keyspace - keyspace :                  0   -->  5950
# info stats    - total_command_processed :  34   --> 31053
#
```
I used Redis `info` command to measure the dimensions above. 
My machine is a 16GB, Intel® Core™ i3 CPU 540 @ 3.07GHz × 4 
Then I cleared completely Redis and repeated the loading of the same dataset but this time a hypergraph was created with the procedure I described you above.
```
# ---------------------------------------- TRIADB ---------------------------------------------------------
# Time    : 55sec
# Stages                                  : CLEAN --> +init --> +dict  --> +loaded data
# info memory   - user_memory_dataset     : 51KB  --> 69KB  --> 103KB  -->        17.7MB (.rdb=5.8MB) ≈ 2.4 x 7
# info keyspace - keyspace :                 0    -->  33   -->  194   -->     34275 ≈ 5950 x 6
# info stats    - total_command_processed :  37   --> 270   --> 6773   --> 1,407,913
#
```
There are several stages before loading the data, and these are shown in my measurements. **Apart from the huge difference in time (55sec vs 2 sec) you will also notice that there is an explosion of the total commands processed (1.4M TRIADB  vs 31K ROM)**. The keyspace and memory differences are reasonable, i.e. I am not simply storing values in a hash and creating indexes, I am also linking bidirectionally each value with a record (row) id. 

**That translates into a 1000 x 18 x 2 = 36000 add link operations and 1000 x 19 = 19000 add node operations.**

But I guess all these operations are taking a huge number of round-trips between Python client and Redis server and If I am not wrong there is no Lua optimization here, neither I found a pipeline mechanism or session batch processing. You also mentioned at some part of ROM manual that ROM `session` is dump, i.e. there isn't a transaction implemented for pipelining commands to the Redis server. I have seen such good ideas implemented in Python stdnet project (see [many-to-many][1], [transaction][2]), e.g. build OneToMany relationship on the server with lua behind the scene. Unfortunately the project has been abandoned and I could not run their current version with Python 3.6

#### Is there any way to drop down loading time significantly for creation of this hypergraph with ROM ManyToMany relationships based on the current ROM version, or should I forget this as it stands now ?

PS: Another detail in my implementation which I forgot to mention is that currently I use_null_session and save every instance when it is created.

[1]: https://pythonhosted.org/python-stdnet/examples/related.html#many-to-many
[2]: https://pythonhosted.org/python-stdnet/api/models.html#transaction
[3]: http://healis.eu/media//triadb/SPC-Hypergraph-References.png

**Comment by josiahcarlson** [comment]
Without seeing your loading code, I can't say why the Triadb method is that much slower. I can make guesses, but I can't say for sure. But it is probably related to why you're executing 1.4 million commands with the Triadb vs. ~31k commands rom does natively.

If you're spending a lot of time looking up data, you can try taking one pass to load your hypernodes, keep a local cache of hypernode_value -> rom_object, then build your hyperedge objects based on those cached hypernode references. Or if you're already doing that... maybe you should email me the code and I can see what you're doing.

Generally, rom isn't very fast during inserts, partly because it has to do 1 round trip on object creation to get a primary id, which it uses when the object is actually saved on a 2nd round trip. I don't think it would impossible (but probably some work) to make it only need 1 round trip, and make it mostly pipeline-able, so multiple inserts could be 1 round-trip, but that's a "change rom" situation.

**Comment by healiseu** [comment]
Thanks, it is a tricky situation, when you load data on a clear system then it's probably better if they are deduplicated first so that saving is not raising unique exception (that is a requirement of TRIADB, it works with single instances for each data type domain, collection). 
In that case there is also no need to look for existing, already saved values in the system. Then my question is whether I am going to reduce significantly round-trips if I use_rom_session() instead of use_null_session() to save all newly created instances.

**Comment by josiahcarlson** [comment]
The session can help in some cases, but overall in this situation you're much better off keeping your own cache. I'm using a variation of https://github.com/josiahcarlson/rom/issues/126 as a starting point to explain, and I'm taking dim2 as being names for attributes and dim3 as being the value for attributes:

```
rom.util.use_null_session()

class IdentityLookup(rom.Model):
    value = rom.Text(unique=True)

class HyperEdge(rom.Model):
    data = rom.OneToMany('ModelTriple')

class ModelTriple(rom.Model):
    edge = rom.ManyToOne('HyperEdge', 'cascade')
    name = rom.ManyToOne('IdentityLookup', 'restrict')
    value = rom.ManyToOne('IdentityLookup', 'restrict')

cache = {}
def get_or_create(value):
    if value not in cache:
        # can skip IdentityLookup.get_by() if this is a completely new load
        v = IdentityLookup.get_by(value=value)
        if not v:
            v = IdentityLookup(value=value)
            v.save()
        cache[value] = v
    return cache[value]

for row in csv.DictReader(...):
    x = HyperEdge()
    x.save()
    for name, value in row.items():
        n = get_or_create(name)
        v = get_or_create(value)
        ModelTriple(edge=x, name=n, value=v).save()
```

**Comment by healiseu** [comment]
Yes, your python set cache might be useful if its size is not too big, otherwise membership check will become expensive. My previous benchmarks were based on TRIADB `add_datum (value)`, `add_item`, `add_row` logic with only `get_by` lookup but for bulk insertion it seems a columnar approach can be proven to be advantageous. For example index selectivity is an important parameter and the execution has to be planned and optimized accordingly. I have to run more tests to measure that.

**Comment by josiahcarlson** [comment]
If you wanted to go to the extreme with column-loading, you could do something like:
`awk -F, '{ print $2 , $0 }' input.csv | sort | col_2_sort.csv`

That at least avoids having to write other column sorting software.

**Comment by healiseu** [comment]
Panda's is always handy for [reading flat files][1] and it can [sort columns][2]

[1]: https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html
[2]: http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.sort_values.html

**Comment by josiahcarlson** [comment]
Numpy is just packed array data with a few bytes of header information. Even regular Python is going to be high memory use compared to using Numpy (40 megs for a list of floats, compared with 8 megs for the an array of doubles in numpy). Also, Redis requires a name for every float (which involves a collection of other related data), unless they are just entries in a list or set (which have other limitations).

There are ways of reducing memory use in Redis, sometimes by a factor of 10x. Most of those techniques are covered in chapter 9 of my book, with later chapters covering writing higher-level structures with Lua that use those memory optimizations.

Part of why I am working on my own fork is because I'm working on solving a variety of existing problems, including some of those memory storage problems (along with a collection of other problems).

**Comment by healiseu** [comment]
Yes, memory storage (data representation) and massive insertion of data are related problems. From the developer's point of view I would expect to load a dataframe (table) of data on a client-server system as fast as possible and if it's a memory data store I would expect it also to fit in a minimum size including indexes. 

You have referred to fast memory data structures like Numpy, recently I started studying such software tools. There is already a good established trend to use Apache Arrow tools for fast in-memory processing and interprocess communication. [Apache Arrow][1] acts as a high-performance interface between various systems.

My work starts at the staging area, i.e. management of data resources, data models, mapping between the two, loading data/indexes in memory and then performing queries (filtering) and aggregations. It would be nice to see if we can link the two projects.

[1]: https://arrow.apache.org/docs/python/index.html

**Comment by josiahcarlson** [comment]
> if it's a memory data store I would expect it also to fit in a minimum size including indexes.

Redis is not good with memory, and your expectations do not coincide with reality.

> Apache Arrow acts as a high-performance interface between various systems.
> It would be nice to see if we can link the two projects.

I don't know what you want. Can you tell me what you want? Do you want to be able to send some numpy object to Redis and have it do something with it? Do you want some sort of shared-memory Redis on your same machine? What are you proposing?

**Comment by healiseu** [comment]
You may find this last comment I wrote [here][1] relevant, there are many puzzles to solve in TRIADB associative - semiotic - hypergraph engine. One of them is fast construction of bitmap secondary indices. I would prefer to demonstrate that to you and others instead of explaining it with words and ask for your valuable assistance when I am in need to solve a very specific implementation task. 

[1]: https://github.com/RedisLabsModules/RediSearch/issues/493#issuecomment-422552591

**Comment by josiahcarlson** [comment]
1. You come to me asking for help, asking questions about how my project works.
2. You tell me how I should build my software, without providing *any* specifics, without any knowledge about how the underlying *anything* works.
3. You provide links to other projects where you're basically doing the same thing, comparing completely unrelated pieces of software, then tell me "read what I wrote over there, you will know what I want"

I don't know what you want because you won't speak plainly about what you want. And If you're not going to speak plainly, and you're not going to answer *my* questions about what you want, then I don't need to help you.

Please stop posting questions on my projects.

**Comment by healiseu** [comment]
Hi, @josiahcarlson I think I owe you a final reply here, since you are complaining in public.
1. Yes if one is interested in an open-source project he asks questions, suggests features to add, etc...
2. That is not true, revisit my github issues/responses. Do whatever you like with your software. I have only tried to use ROM in my project.
3. The links have been added for the specific issue I had with ROM, and Redis in general, about very slow massive insertion and memory allocation.

And finally, If you see the [issues I opened][1], four out of seven have been marked with a `bug` tag. So I guess I helped you this way to improve your software. And I have also given you plenty of feedback through Q&A. You helped me because I helped you too, end of story. Wish you best luck with your new venture.

[1]: https://github.com/josiahcarlson/rom/issues?utf8=%E2%9C%93&q=author%3Ahealiseu

**Comment by josiahcarlson** [comment]
I remember watching at least an hour of video, reading 10 or so pages of docs, visiting a collection of BI software vendors, and then realizing I still had no idea what you wanted. To get even 1 step forward, I had to explicitly ask 6 questions, one of which was, "What do you want from me?"

But yes, you had questions about `rom`, and I answered them. And I answered all the other questions about all the other unrelated technology. And along the way, you hit a couple bugs, and I discovered my own bugs while answering questions. And then you linked to a RediSearch issue. RediSearch is a piece of software that's the closest thing to a direct competitor to `rom` that exists, but with a trash license (RediSearch is better in a lot of ways, and few firms would use both).

Now, you calling all of the issues, support questions about rom, questions about other technology, sending me off to research and collect information to *discover* what you want... as "helping Josiah", and not actually "requesting support for software whose license explicitly states, 'no warranty'"? You didn't submit a PR or commit a line to my repo, you didn't write my docs, and you've not supported any of my users. So you didn't do the work, and yours is the most ridiculous "but I helped" claim I've heard in a long time. You can fuck right the fuck off with that shit.

But I'm not interested (and am not at all incentivized) in helping support software with a trash license, so no helping folks with RediSearch. And I'm not really interested in continuing to try to *guess* at what you want, because that's not how any kind of support, question / answer, development partner, manager / engineer, or even developer / client relationship works. Like at all.

TL;DR: I'm not your therapist, and I shouldn't have to guess or interrogate you to discover what you want. But that has been my experience 100% of the time, and I just don't have the energy.

Future readers: don't claim you did work you didn't, and try to get to the point of your questions / requests.
